src/main/java/soot/jimple/internal/JInterfaceInvokeExpr.java

```diff
@@ -19,54 +19,65 @@
  */
 
 /*
- * Modified by the Sable Research Group and others 1997-1999.  
+ * Modified by the Sable Research Group and others 1997-1999.
  * See the 'credits' file distributed with Soot for the complete list of
  * contributors.  (Soot is distributed at http://www.sable.mcgill.ca/soot)
  */
 
-
-
-
-
-
 package soot.jimple.internal;
 
-import soot.*;
-import soot.jimple.*;
+import soot.SootClass;
+import soot.SootMethod;
+import soot.SootMethodRef;
+import soot.SootMethodRefImpl;
+import soot.Value;
+import soot.ValueBox;
+import soot.jimple.Jimple;
+
+import java.util.ArrayList;
+import java.util.List;
 
-import java.util.*;
+public class JInterfaceInvokeExpr extends AbstractInterfaceInvokeExpr {
 
-public class JInterfaceInvokeExpr extends AbstractInterfaceInvokeExpr
-{	
-    public JInterfaceInvokeExpr(Value base, SootMethodRef methodRef, List<? extends Value> args)
-    {
+    public JInterfaceInvokeExpr(Value base, SootMethodRef methodRef, List<? extends Value> args) {
         super(Jimple.v().newLocalBox(base), methodRef, new ValueBox[args.size()]);
 
         //Check that the method's class is resolved enough
         //CheckLevel returns without doing anything because we can be not 'done' resolving
         methodRef.declaringClass().checkLevelIgnoreResolving(SootClass.HIERARCHY);
+
         //now check if the class is valid
-        if(!methodRef.declaringClass().isInterface() && !methodRef.declaringClass().isPhantom()) {
-        	throw new RuntimeException("Trying to create interface invoke expression for non-interface type: " +
-        			methodRef.declaringClass() +
-        			" Use JVirtualInvokeExpr or JSpecialInvokeExpr instead!");
```

Comment:
Here, you allow methods that aren't declared by interfaces but normal classes as well. Ist this correct? As I understand the language specification, this is illegal. Can you craft an example where an invokeinterface instruction can actually invoke a class declared method?

Reply:
You must be right. I'll remove this commit.
Btw, I haven't found any mentions in the JVM spec that this replacement is impossible. Do yo have this information at hand?

Reply:
https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.10.1.9.invokeinterface

To me, the spec is a bit ambiguous. 

"Its first operand, CP, refers to a constant pool entry denoting an interface method named MethodName with descriptor Descriptor that is a member of an interface MethodIntfName."

I interpret this as "The method has to be an interface method".

However, this would be redundant then:

"MethodName is not <init>.

MethodName is not <clinit>."

The safest way to check this would be crafting an example using an invokeinterface instruction to call a non-interface method and see if it works.

Reply:
To check this I created next class:
```java
public class TestClass {
    public void testMethod() {
        this.toString();
    }
}
```
To call `toString` method I used `invokeinterface`:
```java
Jimple.v().newInterfaceInvokeExpr(body.getThisLocal(), toString.makeRef());
```
When I loaded this class and tried to call `testMethod` I got this exception:
```
Caused by: java.lang.IncompatibleClassChangeError: Found class java.lang.Object, but interface was expected
	at TestClass.testMethod(Unknown Source)
```
So we cannot replace `invokeinterface` with `invokevirtual` without consequences. The check in `soot.jimple.internal.JInterfaceInvokeExpr#JInterfaceInvokeExpr` was correct.

*I reverted this commit and the PR is ready to be merged*

Reply:
Thanks for clarifying!